store/store.go

```diff
@@ -382,16 +511,6 @@ func (bs *BlockStore) SaveBlock(block *types.Block, blockParts *types.PartSet, s
 		panic(err)
 	}
 
-	// Done!
-	bs.mtx.Lock()
-	bs.height = height
-	if bs.base == 0 {
-		bs.base = height
-	}
-	bs.mtx.Unlock()
-
-	// Save new BlockStoreState descriptor. This also flushes the database.
-	bs.saveState()
```

Comment:
Note how this flushes the database. We're not using `SetSync()` above, for performance reasons, which means the OS is just buffering the writes in memory. We need to make sure it's flushed to disk at the end of this method, either by changing the last `Set()` call to `SetSync()` or doing a final explicit sync write.

Reply:
Sure. Just to clarify, databases will also flush stuff to disk when they think the time is right, not just when we specify it's time to flush by calling `SetSync()`?

Reply:
This is mostly managed by the operating system, not the database, but yeah it will flush to disk when appropriate. However, the OS may crash or lost power, in which case it won't make it to disk. This can cause problems, e.g. making a node double-sign if it crashed after submitting a previous vote but didn't persist a record of that vote to disk.